docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+
+* Bad compile-time performance. Due to the way flattening works in Ether and due
+  to a GHC bug, the compile-time performance was devastating. Turning ``-O2``
+  could mean hours of compilation and required up to 65 GIGABYTES of RAM
+  (ridiculous!). This was because GHC generated an exponential amount of
+  coercions (as evidenced by investigating .hi-files). Basically, @int-index no longer
+  can recommend Ether to people as there is no good solution to this known.
+
+Conclusion:
+
+* Migration to Ether allowed us to remove an immense amount of boilerplate,
+  modularize the code, and get good run-time performance. However, lack of
+  run-time configuability was quite inconvenient, and bad compile-time
+  performance marked this approach a no-go.
+
+The Modern Era
+---------------
+
+After we've realized what led to bad compile-time performance, @int-index came up with an
+idea of ``ExecMode``. Basically, we continued to use classes from ``Ether``, but
+rather than having numerous ``IdentityT`` layers there was a single ``newtype``
+wrapper around ``ReaderT ModeEnv Production`` at the bottom. This solved the
+compile-time performance issue completely at the cost of a moderate increase in
+boilerplate.
+
+However, FPComplete began to see ``ether`` as an Enemy of The State, and asked
+us to purge its remains. Now we were supposed to remove all our custom classes,
+replacing them with method records. Those records would go into a ``ReaderT``
+and be passed everywhere manually (as opposed to instance search). Instead of
+distinct ``Ether.MonadReader`` constraints, now we had to use ``MonadReader
+ctx``, passing an annoying ``ctx`` parameter everywhere, and placing constraints
+on it. The final transformer stack is just ``ReaderT ModeCtx Production``, and
+not even a newtype on top.
+
+Technically, we're still in the process of migration, as we haven't removed all
+of our custom classes yet. Just to clarify: in this section we'll discuss the
+current transitional state, and it's more painful than what was actually
+proposed by FPComplete.
+
+
+Also, there is compliсation, that we actually have three aproaches combined:
+	
+* General approach a la ExecMode(s)
+* Usage of reflection (which is in fact custom case for ReaderT for constant data within single execution)
+* Method dictionaries (for supporting SendActions, a primitive from networking)
+* ReaderT with method dictionaries
+
+General approach
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now our code follows this pattern::
+
+    -- effect definitions
+
+    class MonadBaz
+        baz :: ...
+
+    defaultBaz = ...
+
+    class MonadZaz
+        zaz :: ...
+
+    defaultZaz = ...
+
+
+    -- mode definitions
+
+    data QuuxCtx = ...
+
+    type QuuxMode = ReaderT QuuxCtx Production
+
+    instance HasFoo ModeCtx
+    instance HasBar ModeCtx
+
+    instance MonadBaz QuuxMode
+        baz = defaultBaz
+
+    instance MonadZaz QuuxMode
+        zaz = defaultZaz
+
+
+Pros:
```

Comment:
Also simplicity, which is really important.

Reply:
I don't think it's particularly simple, at least as long as we have custom classes with `ReaderT` instances.

Reply:
IDK, i think this approach is the most simple of all possible.